### ISSUE TITLE
StrongSchwazBC compile error

### ISSUE THREAD

**Issue by ibaned** [gahansen/Albany#207]
```
/.../Albany/src/LCM/evaluators/bc/StrongSchwarzBC_Def.hpp(747): error: no operator "<" matches these operands
```
http://cdash.sandia.gov/CDash-2-3-0/viewBuildError.php?buildid=62035

**Comment by ibaned** [comment]
This seems to only fail with Intel compilers.

**Comment by ikalash** [comment]
If you change "auto" in lines 743 and 747 to "int" does it fix the problem?

**Comment by ibaned** [comment]
probably. I don't have an Intel build handy to try

**Comment by ikalash** [comment]
Me neither.  My guess is the "auto" is the problem.  I will go ahead and change it, and we can see what happens with the CDash tomorrow.

**Comment by bartgol** [comment]
This is odd, since, according to the standard [§2.13.2],

"The type of an integer literal is the first of the corresponding list in Table 5 in which its value can be represented."

Table 5 lists `int, long int, long long int`. So `auto i = 0` *should* be equivalent to `int i=0`, and it would be surprising that intel compiler don't do the right thing there. On the other hand, I don't see any other reason why the compiler should complain...

**Comment by ikalash** [comment]
Right...  the only thing that seems to me could be wrong is the "auto" but it seems that the "auto" should work...

**Comment by ibaned** [comment]
Actually, the first line of the compile error text indicates the real problem:
```
operand types are: int < const std::initializer_list
```
So in fact it is `spatial_dimension` that is wrong, either because we are initializing it incorrectly, or because the Intel compiler isn't doing C++11 correctly. I think I've seen the Intel compiler misinterpret brace initialization before.
Unfortunately this means changing the `auto` won't fix it, but probably changing `spatial_dimension` from brace-initialization to older syntax like just `=` would fix it.

**Comment by ibaned** [comment]
@lxmota I think I've seen this pattern in your code a lot. Until the Intel compiler fixes this bug, you've got to be careful not to use brace-initialization too much.

**Comment by ikalash** [comment]
I actually changed already the type of spatial_dimension to "int" in  my checking earlier this morning.

**Comment by ibaned** [comment]
Oh okay. that will probably fix it then. I guess `auto x{some_init_value};` is the dangerous pattern.

**Comment by bartgol** [comment]
Ah, I didn't know the original code had `auto` for the brace-initialized int. In that case, the compiler was right. The following do two different things:

`auto a{0};
auto b = 0;`

This is indeed the *only* case (probably) where `=` initialization is to be preferred to brace initialization. The type deduced by `auto` is `initializer_list` in case of brace initialization.

**Comment by ibaned** [comment]
Apparently this is a point of contention all the way back to the C++ standard:

https://stackoverflow.com/questions/31310167/auto-and-brace-initialization-in-c11-c14

The Intel compiler is actually doing what C++11 says should be done. There is a whole proposed change for C++17 to "fix" this and make it work the way @lxmota expected:

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n3922.html

Long story short: the type should be `std::initializer_list` in the C++11 standard, but in the C++17 standard this will deduce `int` the way we expect. Until C++17 is the norm and that mixup gets resolved, don't combine `auto` and brace initializers, only use one or the other at a time.
